AudioVideoRendererRemote should only serialise TrackInfo if it changes

jyavenard opened the pull request:
#### 7f582bce96ed45d1a34d2061696482b8bcf2fd8d
<pre>
AudioVideoRendererRemote should only serialise TrackInfo if it changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=303433">https://bugs.webkit.org/show_bug.cgi?id=303433</a>
<a href="https://rdar.apple.com/165725525">rdar://165725525</a>

Reviewed by Youenn Fablet.

The MediaSamplesBlock&apos;s TrackInfo object was fully serialised/deserialised
when sent over IPC, even when it didn&apos;t change.
This conversion was one of the top CPU&apos;s ms/cpu usage in the web content process
and caused unnecessary elevated power usage.

We add a new MediaSampleConverter utility class that will track if the metadata (init segment)
of a MediaSample has changed, and if not will re-use the previously converted TrackInfo instead.
A new IPC newTrackInfoForTrack to RemoteAudioVideoRendererProxyManager is added so that we can
send separately the TrackInfo object when it has changed.
MediaSamplesBlock will have their TrackInfo set to null before being set to
the GPU process so that upon receiving we can re-use the existing, previously received,
 TrackInfo.

No change in observable behaviour, covered by existing tests.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/MediaSamplesBlock.cpp:
(WebCore::MediaSamplesBlock::toMediaSample const):
* Source/WebCore/platform/MediaSamplesBlock.h:
* Source/WebCore/platform/graphics/MediaSampleConverter.cpp: Added.
(WebCore::hasSameInitSegment):
(WebCore::MediaSampleConverter::convert):
(WebCore::MediaSampleConverter::hasFormatChanged):
(WebCore::MediaSampleConverter::currentTrackInfo const):
* Source/WebCore/platform/graphics/MediaSampleConverter.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::samplesBlockFromCMSampleBuffer):
* Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
(WebKit::RemoteAudioVideoRendererProxyManager::converterContextFor):
(WebKit::RemoteAudioVideoRendererProxyManager::newTrackInfoForTrack):
(WebKit::RemoteAudioVideoRendererProxyManager::enqueueSample):
* Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.h:
* Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.messages.in:
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
(WebKit::AudioVideoRendererRemote::enqueueSample):
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.h:

Canonical link: <a href="https://commits.webkit.org/304579@main">https://commits.webkit.org/304579@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/13b91a678841324e349ec3a488ced5111c3b5c81

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136031 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8383 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47305 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143730 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4d936e52-cfe9-4893-ab6b-b4a5ef64e52f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137900 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9052 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8230 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103952 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/eeef3444-144b-44e6-87a5-dce121a8c49b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138977 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6566 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121898 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84831 "Found 1 new API test failure: WPE/TestWebKitWebView:/webkit/WebKitWebView/save (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a8933bae-7ac4-42a9-a697-2f0f38a1bd80) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/3900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4331 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/115524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/40093 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146483 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8066 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40660 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/112307 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8082 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6769 "Build is in progress. Recent messages:OS: Tahoe (26.2), Xcode: 26.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; Running layout-tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/112701 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28592 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6157 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118202 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8116 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/36268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7832 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71673 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8053 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/7912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

youennf reviewed (commented):
Could you describe why this change is needed? Is it just an optimisation?

youennf commented on Source/WebCore/platform/graphics/MediaSampleConverter.cpp:
```diff
@@ -0,0 +1,88 @@
+    return true;
+}
+
+bool MediaSampleConverter::hasSameInitSegment(const MediaSample& sampleA, const MediaSample& sampleB)
```
Can it be a free function and not declared in the header?

youennf commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -250,16 +250,32 @@ void RemoteAudioVideoRendererProxyManager::requestMediaDataWhenReady(RemoteAudio
+    auto& converterContext = converterContextFor(iterator->value, trackIdentifier);
+
+    samplesBlock.setInfo(converterContext.first.get());
+    if (RefPtr mediaSample = converterContext.second.convert(WTFMove(samplesBlock))) {
```
Should we ASSERT(mediaSample)?

jyavenard replied:
the code handles when the conversion failed and a nullptr was returned.
The conversion is fallible if it's running out of memory for example

youennf commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -250,16 +250,32 @@ void RemoteAudioVideoRendererProxyManager::requestMediaDataWhenReady(RemoteAudio
+
+    auto& converterContext = converterContextFor(iterator->value, trackIdentifier);
+
+    samplesBlock.setInfo(converterContext.first.get());
```
Should we ASSERT(converterContext.first)?

jyavenard replied:
I'll even make it a `MESSAGE_CHECK_COMPLETION` one.

youennf commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -250,16 +250,32 @@ void RemoteAudioVideoRendererProxyManager::requestMediaDataWhenReady(RemoteAudio
+    auto iterator = m_renderers.find(identifier);
+    MESSAGE_CHECK_COMPLETION(iterator != m_renderers.end(), completionHandler(false));
+
+    auto& converterContext = converterContextFor(iterator->value, trackIdentifier);
```
Using references like this is not really aligned with safer cpp guidelines.

jyavenard replied:
The object isn't recounted. How is that not safe here?

youennf commented on Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
```diff
@@ -475,7 +475,17 @@ void AudioVideoRendererRemote::enqueueSample(TrackIdentifier trackIdentifier, Re
+        RefPtr trackInfo = block->info();
+        if (formatChanged)
+            connection.send(Messages::RemoteAudioVideoRendererProxyManager::NewTrackInfoForTrack(renderer.m_identifier, trackIdentifier, Ref { const_cast<WebCore::TrackInfo&>(*trackInfo) }), 0);
+        block->setInfo(nullptr);
```
Do you plan a refactoring to not have to do that nullptr assignment?

jyavenard replied:
What improvement would that bring?

jyavenard replied:
I'll make the converter return such MediaSampleBlock to start so it's no longer up to the caller to set it to null

jyavenard committed:
AudioVideoRendererRemote should only serialise TrackInfo if it changes
https://bugs.webkit.org/show_bug.cgi?id=303433
rdar://165725525

Reviewed by Youenn Fablet.

The MediaSamplesBlock's TrackInfo object was fully serialised/deserialised
when sent over IPC, even when it didn't change.
This conversion was one of the top CPU's ms/cpu usage in the web content process
and caused unnecessary elevated power usage.

We add a new MediaSampleConverter utility class that will track if the metadata (init segment)
of a MediaSample has changed, and if not will re-use the previously converted TrackInfo instead.
A new IPC newTrackInfoForTrack to RemoteAudioVideoRendererProxyManager is added so that we can
send separately the TrackInfo object when it has changed.
MediaSamplesBlock will have their TrackInfo set to null before being set to
the GPU process so that upon receiving we can re-use the existing, previously received,
 TrackInfo.

No change in observable behaviour, covered by existing tests.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/MediaSamplesBlock.cpp:
(WebCore::MediaSamplesBlock::toMediaSample const):
* Source/WebCore/platform/MediaSamplesBlock.h:
* Source/WebCore/platform/graphics/MediaSampleConverter.cpp: Added.
(WebCore::hasSameInitSegment):
(WebCore::MediaSampleConverter::convert):
(WebCore::MediaSampleConverter::hasFormatChanged):
(WebCore::MediaSampleConverter::currentTrackInfo const):
* Source/WebCore/platform/graphics/MediaSampleConverter.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::samplesBlockFromCMSampleBuffer):
* Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
(WebKit::RemoteAudioVideoRendererProxyManager::converterContextFor):
(WebKit::RemoteAudioVideoRendererProxyManager::newTrackInfoForTrack):
(WebKit::RemoteAudioVideoRendererProxyManager::enqueueSample):
* Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.h:
* Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.messages.in:
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
(WebKit::AudioVideoRendererRemote::enqueueSample):
* Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.h:

Canonical link: https://commits.webkit.org/304579@main
